### ISSUE TITLE
Wrong placement of @NonNull annotation when referencing inner class / enum

### ISSUE THREAD

**Issue by ChristianCiach** [google/auto#446]
Simple example:

```
@NonNullByDefault
public class MyOuterClass {

	public enum MyEnum {
		A, B;
	}

	@AutoValue
	static abstract class MyInnerClass {

		public static MyInnerClass of(MyEnum myEnum) {
			return new AutoValue_MyOuterClass_MyInnerClass(myEnum);
		}

		public abstract MyEnum myEnum();
	}
}
```

This generates a compile error inside `AutoValue_MyOuterClass_MyInnerClass` that says:

> "Type annotations are not allowed on type names used to access static members"

The reason is the wrong placement of the "@NonNull" annotation. See this generated snipped by AutoValue:

```
  @NonNull
  @Override
  public MyOuterClass.MyEnum myEnum() {
    return myEnum;
  }
```

In this case, "@NonNull" refers to "MyOuterClass". It's suprising and ugly, but the correct placement of the annotation would be this:

```
  @Override
  public MyOuterClass.@NonNull MyEnum myEnum() {
    return myEnum;
  }
```

Following [the Java Language Specification, Chapter 9.7.4](http://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.7.4) it probably would be better to always write type annotations immediately before the type to which they apply.

> It is customary, though not required, to write declaration annotations before all other modifiers, and type annotations immediately before the type to which they apply. 

This way the annotation can be seen as part of the type, which is close enough to the truth.

**Comment by eamonnmcmanus** [comment]
Thanks! We're aware that the handling of type-targeted annotations needs work, and this is a good example.

**Comment by ChristianCiach** [comment]
With the update from 1.3 to 1.4.1 I now have this error on generated classes where this error previously didn't occure.

**Comment by eamonnmcmanus** [comment]
Yes. Previously, in an example like the above, the generated code would import MyOuterClass.MyEnum and reference it as just MyEnum. Now it always references nested classes starting from the outermost class, so you get `@Nonnull MyOuterClass.MyEnum` instead of `MyOuterClass.@Nonnull MyEnum`. This needs to be fixed.

**Comment by eamonnmcmanus** [comment]
This is fixed in 1.4-SNAPSHOT and the fix will be in 1.5.
